### PR TITLE
Multithreading bug with accessing this.searchProviders

### PR DESCRIPTION
**Pull request by fpichler**
synchronise access to plugin instance variable searchProviders to solve a multithreading issue.

Please review and merge if ok.

### CODE REVIEW

File: src/main/java/com/xpn/xwiki/plugin/lucene/IndexUpdater.java [hunk 1]
```diff
@@ -247,7 +247,7 @@ private void updateIndex()
                 }
             }
 
-            this.plugin.openSearchers(context);
+            this.plugin.openSearchers();
```
Comment (msladek) [inline src/main/java/com/xpn/xwiki/plugin/lucene/IndexUpdater.java]: Compile error, context is still a parameter of openSearcher

File: src/main/java/com/xpn/xwiki/plugin/lucene/LucenePlugin.java [hunk 2]
```diff
@@ -864,20 +859,33 @@ public void flushCache(XWikiContext context) {
...
+          this.searcherProvider.markToClose();
+          this.searcherProvider = null;
+        }
+        this.searcherProvider = getSearcherProviderManager().createSearchProvider(
+            createSearchers(this.indexDirs));
+      } catch (Exception exp) {
+        LOGGER.error("Error opening searchers for index dirs [{}]", context.getWiki(
+            ).Param(PROP_INDEX_DIR), exp);
+        throw new RuntimeException("Error opening searchers for index dirs "
+            + context.getWiki().Param(PROP_INDEX_DIR), exp);
```
Comment (msladek) [inline src/main/java/com/xpn/xwiki/plugin/lucene/LucenePlugin.java:889]: Inteded to also wrap non-RuntimeExceptions (IOException) into RuntimeException?

**Comment by msladek** [comment]
If i'm not mistaken, this still is not a fix:

Between the synchroized "getSearchProvider()"-call in "getSearchResults...()" and "theSearcherProvider.connect()" on line #530, an outside call of "openSearchers()" can still mark the provider closed (and connect will therefore fail).

**Comment by fpichler** [comment]
Thanks for you hint.
Corrected, please check again.

**Comment by msladek** [comment]
looks good now

**Pull request merged**